File: quickSort.py
def get_count(arr, s, e, pivot):
    count = 0

    for i in range(s, e):
        if arr[i] < pivot:
            count += 1

    return count

def partition(arr, s, e):
    pivot = arr[s]
    count = get_count(arr, s, e, pivot)
    pivot_index = s + count

    arr[pivot_index], arr[s] = arr[s], arr[pivot_index]

    i, j = s, e

    while i < pivot_index and j > pivot_index:
        while arr[i] < arr[pivot_index]:
            i += 1

        while arr[j] > arr[pivot_index]:
            j -= 1

        if i < pivot_index and j > pivot_index:
            arr[i], arr[j] = arr[j], arr[i]
            i += 1
            j -= 1

    return pivot_index

def quick_sort(arr, s, e):
    if s >= e:
        return

    p = partition(arr, s, e)

    quick_sort(arr, s, p - 1)
    quick_sort(arr, p + 1, e)

if __name__ == "__main__":
    arr = [234, 432, 234, 231, 12, 432, 1324, 5432, 234, 453]
    
    s, e = 0, len(arr) - 1

    quick_sort(arr, s, e)

    for i in range(len(arr)):
        print(arr[i], end=" ")

File: QuickSort.cpp
#include <iostream>
using namespace std;

int getCount(int arr[],int s, int e,  int pivot) {
    int count = 0;
    
    for(int i = s; i < e; i++) {
        if (arr[i] < pivot)
            count++;
    }

    return count;

}

int partition(int *arr, int s, int e) {
    // pivot ele
    int pivot = arr[s];

    // number of elements smaller than pivot element
    int count = getCount(arr,s, e, pivot);

    // correct position of pivot index
    int pivotIndex = s + count;

    swap(arr[pivotIndex], arr[s]);

    int i = s, j = e;

    while (i < pivotIndex && j > pivotIndex) {
        if (arr[i] < arr[pivotIndex]){
            i++;
        }
        
        else if (arr[j] > arr[pivotIndex]) {
            j--;
        }

        if (arr[i] > arr[pivotIndex] && arr[j] < arr[pivotIndex]) 
            swap(arr[i++], arr[j--]);
    }

    return pivotIndex;

}

void quickSort(int* arr, int s, int e) {
    if (s >= e){
        return;
    }

    int p = partition(arr, s, e);

    quickSort(arr, s, p - 1);
    quickSort(arr, p + 1, e);
}

int main () {

    int arr[10] = {234,432,234,231,12,432,1324,5432,234,453};
    
    int s = 0, e = 10 - 1;

    quickSort(arr,s , e );

    for(int i = 0; i < 10; i++){
        cout << arr[i] << " ";
    }

    return 0;
}
